Mathlib/NumberTheory/NumberField/Completion.lean

```diff
@@ -0,0 +1,288 @@
+
+variable {L : Type*} [NormedField L] [CompleteSpace L] {f : WithAbs v →+* L} {v}
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field
+`L`, then we can extend that embedding to an embedding on the completion `v.completion →+* L`. -/
+def extensionEmbedding_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    v.completion →+* L :=
+  UniformSpace.Completion.extensionHom _
+    (WithAbs.isUniformInducing_of_comp h).uniformContinuous.continuous
+
+theorem extensionEmbedding_of_comp_coe
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) (x : K) :
+    extensionEmbedding_of_comp h x = f x := by
+  rw [← UniformSpace.Completion.extensionHom_coe f
+    (WithAbs.isUniformInducing_of_comp h).uniformContinuous.continuous]
+  rfl
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* L` preserves distances. -/
+theorem extensionEmbedding_dist_eq_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective)
+    (x y : v.completion) :
+    dist (extensionEmbedding_of_comp h x) (extensionEmbedding_of_comp h y) =
+      dist x y := by
+  refine (UniformSpace.Completion.induction_on₂ x y ?_ (fun x y => ?_))
+  · refine isClosed_eq ?_ continuous_dist
+    exact (continuous_iff_continuous_dist.1 (UniformSpace.Completion.continuous_extension))
+  · simp only [extensionEmbedding_of_comp_coe]
+    exact UniformSpace.Completion.dist_eq x y ▸ Isometry.dist_eq (WithAbs.isometry_of_comp h) _ _
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* L` is an isometry. -/
+theorem isometry_extensionEmbedding_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    Isometry (extensionEmbedding_of_comp h) :=
+  Isometry.of_dist_eq <| extensionEmbedding_dist_eq_of_comp h
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* L` is a closed embedding. -/
+theorem isClosedEmbedding_extensionEmbedding_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    IsClosedEmbedding (extensionEmbedding_of_comp h) :=
+  (isometry_extensionEmbedding_of_comp h).isClosedEmbedding
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field
+that is locally compact, then the completion of the first normed field is also locally compact. -/
+theorem locallyCompactSpace [LocallyCompactSpace L]
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective)  :
+    LocallyCompactSpace (v.completion) :=
+  (isClosedEmbedding_extensionEmbedding_of_comp h).locallyCompactSpace
+
+end AbsoluteValue.Completion
```

Comment:
Everything until here is unrelated to number fields, right? Is there a better place?

Reply:
Yes there probably is. They could go in the `WithAbs.lean` file maybe? Or in [`UniformSpace.AbsoluteValue`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/AbsoluteValue.html)? I think `WithAbs.lean` might be a better choice, but let me know if you think otherwise.

Reply:
It seems reasonable also to me. (If you want to open a separate PR we will not complain...)